### ISSUE TITLE
Sends multiple connection control packet throw error [MQTT-3.1.0-2]

### ISSUE THREAD

**Issue by freeseawind** [vert-x3/vertx-mqtt#33]
The client sends multiple connection control packet, server through follow exception:
Unhandled exception java.lang.IllegalArgumentException: Duplicate handler name: idle

May be add the following code in MqttConnection.java

  synchronized void handleMessage(Object msg) {
   ..........
      switch (mqttMessage.fixedHeader().messageType()) {

        case CONNECT:
          
         if(endpoint == null || !endpointisConnected())
         {
                .......
         }
 }

**Comment by dachaac** [comment]
http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718028:
"A Client can only send the CONNECT Packet once over a Network Connection. The Server MUST process a second CONNECT Packet sent from a Client as a protocol violation and disconnect the Client [MQTT-3.1.0-2]"

But I believe there should not be exception thrown in this case.

**Comment by Sammers21** [comment]
I have just reproduced the error. Here is a full log with stacktraces:
```
[MQTT SERVER] MQTT server is listening on port 1883
[MQTT CLIENT] first CONNECT sent
[MQTT SERVER] MQTT client [966d72f2-f86d-4e00-9c8a-03418b748f30] request to connect, clean session = true
[MQTT SERVER] [will topic = null msg = null QoS = 0 isRetain = false]
[MQTT SERVER] [keep alive timeout = 100]
[MQTT CLIENT] second CONNECT sent
июл 31, 2017 9:13:46 PM io.vertx.core.impl.ContextImpl
SEVERE: Unhandled exception
java.util.NoSuchElementException: timeoutOnConnect
	at io.netty.channel.DefaultChannelPipeline.getContextOrDie(DefaultChannelPipeline.java:1080)
	at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:434)
	at io.vertx.mqtt.impl.MqttConnection.handleConnect(MqttConnection.java:227)
	at io.vertx.mqtt.impl.MqttConnection.handleMessage(MqttConnection.java:95)
	at io.vertx.mqtt.impl.MqttServerImpl.lambda$null$3(MqttServerImpl.java:89)
	at io.vertx.core.net.impl.NetSocketImpl.handleMessageReceived(NetSocketImpl.java:358)
	at io.vertx.core.net.impl.NetServerImpl$2.handleMessage(NetServerImpl.java:423)
	at io.vertx.core.net.impl.NetServerImpl$2.handleMessage(NetServerImpl.java:420)
	at io.vertx.core.net.impl.VertxHandler.lambda$channelRead$1(VertxHandler.java:150)
	at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:343)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:201)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:148)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:287)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:293)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:267)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:129)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:642)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:565)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:479)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:441)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:748)

[MQTT CLIENT] third CONNECT sent
июл 31, 2017 9:13:47 PM io.vertx.core.impl.ContextImpl
SEVERE: Unhandled exception
java.util.NoSuchElementException: idle
	at io.netty.channel.DefaultChannelPipeline.getContextOrDie(DefaultChannelPipeline.java:1080)
	at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:434)
	at io.vertx.mqtt.impl.MqttConnection.handleConnect(MqttConnection.java:226)
	at io.vertx.mqtt.impl.MqttConnection.handleMessage(MqttConnection.java:95)
	at io.vertx.mqtt.impl.MqttServerImpl.lambda$null$3(MqttServerImpl.java:89)
	at io.vertx.core.net.impl.NetSocketImpl.handleMessageReceived(NetSocketImpl.java:358)
	at io.vertx.core.net.impl.NetServerImpl$2.handleMessage(NetServerImpl.java:423)
	at io.vertx.core.net.impl.NetServerImpl$2.handleMessage(NetServerImpl.java:420)
	at io.vertx.core.net.impl.VertxHandler.lambda$channelRead$1(VertxHandler.java:150)
	at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:343)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:201)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:148)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:293)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:267)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:129)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:642)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:565)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:479)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:441)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:748)
```
The reproducer can be found [here](https://github.com/Sammers21/vertx-mqtt-client-example/tree/reporoduce_vertx_mqtt_server_problem).

Raw log: https://pastebin.com/raw/CSSiv5zA

**Comment by Sammers21** [comment]
There are two exceptions: *java.util.NoSuchElementException: idle* and *java.util.NoSuchElementException: timeoutOnConnect*.

So, it seems that @freeseawind faced a different from these two ones.

@freeseawind can you provide a reproducer?

**Comment by freeseawind** [comment]
vertx-mqtt not found in the maven repository

**Comment by Sammers21** [comment]
@freeseawind you are right, there is no such artifact in maven central, but you can build artifact locally in this way:
```bash
git clone https://github.com/vert-x3/vertx-mqtt && cd vertx-mqtt && mvn clean install
```

So, with reproducer, I provided the exception is not java.lang.IllegalArgumentException right? If so can you provide the full log(something like this https://pastebin.com/raw/CSSiv5zA)? It would be very useful.

**Comment by freeseawind** [comment]
八月 01, 2017 2:32:14 下午 io.vertx.core.impl.ContextImpl
严重: Unhandled exception
java.lang.IllegalArgumentException: Duplicate handler name: idle
	at io.netty.channel.DefaultChannelPipeline.checkDuplicateName(DefaultChannelPipeline.java:1062)
	at io.netty.channel.DefaultChannelPipeline.filterName(DefaultChannelPipeline.java:293)
	at io.netty.channel.DefaultChannelPipeline.addBefore(DefaultChannelPipeline.java:250)
	at io.netty.channel.DefaultChannelPipeline.addBefore(DefaultChannelPipeline.java:240)
	at io.vertx.mqtt.impl.MqttConnection.handleConnect(MqttConnection.java:275)
	at io.vertx.mqtt.impl.MqttConnection.handleMessage(MqttConnection.java:133)
	at io.vertx.mqtt.impl.MqttServerImpl.handleMsgReceived(MqttServerImpl.java:107)
	at io.vertx.mqtt.impl.MqttServerImpl.handleMsgReceived(MqttServerImpl.java:1)
	at io.vertx.core.net.impl.NetServerBase$ServerHandler.handleMsgReceived(NetServerBase.java:308)
	at io.vertx.core.net.impl.VertxNetHandler.lambda$channelRead$0(VertxNetHandler.java:63)
	at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:335)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:193)
	at io.vertx.core.net.impl.VertxNetHandler.channelRead(VertxNetHandler.java:63)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:122)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:287)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:293)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:267)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1228)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1039)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:411)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:248)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:129)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:642)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:565)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:479)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:441)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:745)

**Comment by freeseawind** [comment]
Here is full log with stack traces https://pastebin.com/raw/htLrvHTj

**Comment by Sammers21** [comment]
@freeseawind, it seems that you are using vertx-mqtt-server:3.4.2. I even found the line where error occurs: https://github.com/vert-x3/vertx-mqtt/blob/9d46fa89ab9181c728e565724390f3c09e5525c7/src/main/java/io/vertx/mqtt/impl/MqttConnection.java#L232 

Such situation will not happen in vertx-mqtt:3.5.0.Beta1, but there are any other issues which can happen like *java.util.NoSuchElementException: idle* and *java.util.NoSuchElementException: timeoutOnConnect* . To fix them i have created PR #43 .

And my recomendation for you just is just wait for vertx-mqtt:3.5.0.Beta1 release which will be attempted to be released in a week after 04/08/2017(I hope my PR will be merged before) or you can set up artifact locally from this repository when it will be merged.

**Comment by ppatierno** [comment]
It should be fixed by 322f9e3b756b28d2335ceff053898ef1dbcce56b
